open-formulieren/open-forms-sdk
# Upgrade to React 18 in dependencies

The React component libraries for NL DS make use of `React.useId` which is React 18+.

**Analysis whether to upgrade or not**

Alternatives explored:

* get the component libraries to distribute individual components instead of a bundle ( :red_circle: lot of work in build tooling)
* (temporarily) use something else for the `React.useId` usage (:red_circle: not desirable - what is the policy on supporting old versions etc.)

Risks:

* Unsure how this plays with potential breaking changes and old API usage in react-formio
* React.StrictMode + our `useEffect` usage may reveal a number of bugs
* Effect on consumers of our NPM package -> should be minimal, as we only use the React API's internally and otherwise there are no breaking changes

-> upgrade to React 18 turns out to be okay-ish, bigger upgrade was react-router from v5 -> v6 but that all seems to work as expected.

**Manual tests performed**

* Configured a form with two steps, started it from a clean session, filled out step 1, step 2, submitted + verified I could download the confirmation PDF
* Added an appointment step to the form (step 3) using the demo plugin, filled out the appoinment details, submitted. In the admin, clicked the "cancel appointment" link and verified it redirects back to the frontend correctly (see also open-formulieren/open-forms#3115 for that). Cancelled appointment & correctly got confirmation screen.
* Clicked through storybook - there are some router-related stories that are broken at the moment (Private API / Form Step).

I did not test the "new" appointment stuff since that's very much in Flux and known to be broken-ish right now, but this is not in production so it's acceptable.

**TODO**

- [x] Fix `Private API / FormStep` stories
